Realm/ObjectServerTests/SwiftFlexibleSyncServerTests.swift

```diff
@@ -1315,6 +1356,318 @@ extension SwiftFlexibleSyncServerTests {
+        Task {
+            print("start task")
+            _ = try await realm.objects(SwiftPerson.self).where { $0.age >= 8 }.subscribe()
+            _ = try await results0.subscribe()
+            XCTAssertEqual(realm.subscriptions.count, 1)
+            ex.fulfill()
+        }
+        print("start await")
+        await fulfillment(of: [ex], timeout: 5.0)
+        XCTAssertEqual(realm.subscriptions.count, 1)
+
+    }
+
+    @MainActor
+    func testSubscribeSameQuerySameName() async throws {
+        try await populateSwiftPerson()
+        let realm = try openFlexibleSyncRealm()
+
+        let results0 = try await realm.objects(SwiftPerson.self).where { $0.age >= 8 }.subscribe(name: "8 or older")
+        realm.syncSession!.suspend()
+        let ex = XCTestExpectation(description: "no attempt to re-create subscription, returns immediately")
+        Task {
+            _ = try await realm.objects(SwiftPerson.self).where { $0.age >= 8 }.subscribe(name: "8 or older")
+            _ = try await results0.subscribe(name: "8 or older")
+            XCTAssertEqual(realm.subscriptions.count, 1)
+            ex.fulfill()
+        }
+        await fulfillment(of: [ex], timeout: 5.0)
+        XCTAssertEqual(realm.subscriptions.count, 1)
+    }
+
+    @MainActor
+    func testSubscribeSameQueryDifferentName() async throws {
+        try await populateSwiftPerson()
+        let realm = try openFlexibleSyncRealm()
+
+        let results0 = try await realm.objects(SwiftPerson.self).where { $0.age >= 8 }.subscribe()
+        _ = try await realm.objects(SwiftPerson.self).where { $0.age >= 8 }.subscribe(name: "8 or older")
+        _ = try await results0.subscribe(name: "older than 7")
+        XCTAssertEqual(realm.subscriptions.count, 3)
+        let first = realm.subscriptions.first(ofType: SwiftPerson.self) { $0.age >= 8 }
+        XCTAssertNil(first?.name)
+        let second = realm.subscriptions[1]
+        XCTAssertEqual(second!.name, "8 or older")
+        let third = realm.subscriptions[2]
+        XCTAssertEqual(third!.name, "older than 7")
+    }
+
+    @MainActor
+    func testSubscribeDifferentQuerySameName() async throws {
+        try await populateSwiftPerson()
+        let realm = try openFlexibleSyncRealm()
+
+        _ = try await realm.objects(SwiftPerson.self).where { $0.age > 8 }.subscribe(name: "group1")
+        _ = try await realm.objects(SwiftPerson.self).where { $0.age > 5 }.subscribe(name: "group1")
+        XCTAssertEqual(realm.subscriptions.count, 1)
+        for subscription in realm.subscriptions {
+            XCTAssertEqual(subscription.queryString, "age > 5")
+        }
+    }
+
+    @MainActor
+    func testUnsubscribe() async throws {
+        try await populateSwiftPerson()
+        let realm = try openFlexibleSyncRealm()
+
+        let results1 = try await realm.objects(SwiftPerson.self).where { $0.lastName == "lastname_3" }.subscribe()
+        XCTAssertEqual(realm.subscriptions.count, 1)
```

Comment:
A more complete test will check the results data after the unsubscription

Reply:
It would. But unsubscribe() doesn't wait for any data to leave the phone. It's synchronous and returns when the subscriptionSet changes are committed locally. As far as I can tell, waitForDownloads and waitForUploads don't cover the case for data _leaving_ the client. And I couldn't figure a way to do that. The only way I could thnk was an arbitrary wait time. Is there a better way?